src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/waypoints_pending_resumption_handler.cc

```diff
@@ -139,71 +144,61 @@ void WayPointsPendingResumptionHandler::on_event(
-    return;
-  }
-  pending_request = pending_requests_[corr_id];
-  pending_requests_.erase(corr_id);
-  if (app_ids_.empty()) {
-    SDL_LOG_ERROR("app_ids is empty");
-    return;
-  }
-  uint32_t app_id = app_ids_.front();
-  app_ids_.pop();
-  auto app = application_manager_.application(app_id);
+  auto current_pending = pending_requests_.front();
+  pending_requests_.pop_front();
+
+  auto app = application_manager_.application(current_pending.app_id_);
   if (!app) {
-    SDL_LOG_ERROR("Application not found " << app_id);
+    SDL_LOG_DEBUG("Application not found " << current_pending.app_id_);
     return;
   }
 
   if (resumption::IsResponseSuccessful(response)) {
-    SDL_LOG_DEBUG("Resumption of subscriptions is successful");
-
+    SDL_LOG_DEBUG("Resumption is successful");
     application_manager_.SubscribeAppForWayPoints(app);
+  }
+  ProcessNextPendingResumption(response);
+}
 
-    for (auto& frozen_resumption : frozen_resumptions_) {
-      auto corr_id = frozen_resumption.request_to_send_
-                         .message[strings::params][strings::correlation_id]
-                         .asInt();
-      RaiseFakeSuccessfulResponse(response, corr_id);
-      application_manager_.SubscribeAppForWayPoints(frozen_resumption.app_id);
-    }
-    frozen_resumptions_.clear();
-  } else {
-    SDL_LOG_DEBUG("Resumption of subscriptions is NOT successful");
-
-    if (frozen_resumptions_.empty()) {
-      SDL_LOG_DEBUG("frozen resumptions list is empty");
-      return;
-    }
-
-    ResumptionAwaitingHandling frozen_resumption = frozen_resumptions_.back();
-    frozen_resumptions_.pop_back();
-    auto resumption_req = frozen_resumption.request_to_send_;
-    const uint32_t cid =
-        resumption_req.message[strings::params][strings::correlation_id]
-            .asInt();
-    const hmi_apis::FunctionID::eType fid =
-        static_cast<hmi_apis::FunctionID::eType>(
-            resumption_req.message[strings::params][strings::function_id]
-                .asInt());
-    subscribe_on_event(fid, cid);
-    auto request =
-        std::make_shared<smart_objects::SmartObject>(resumption_req.message);
-    SDL_LOG_DEBUG("Subscribing for event with function id: "
-                  << fid << " correlation id: " << cid);
-    pending_requests_[cid] = *request;
-    SDL_LOG_DEBUG("Sending request with fid: " << fid << " and cid: " << cid);
-    application_manager_.GetRPCService().ManageHMICommand(request);
+void WayPointsPendingResumptionHandler::ProcessNextPendingResumption(
+    const smart_objects::SmartObject& response) {
+  SDL_LOG_AUTO_TRACE();
+  if (pending_requests_.empty()) {
+    SDL_LOG_DEBUG("No more pending resumptions");
+    return;
   }
+  auto& pending = pending_requests_.front();
+  if (pending.waiting_for_hmi_response_) {
+    SDL_LOG_DEBUG("Request was already sent to HMI for " << pending.app_id_);
+    return;
+  }
+
+  if (!resumption::IsResponseSuccessful(response)) {
```

Comment:
Why do we need to analyze the response. Can we use `application_manager_->IsAnyAppSubscribedForWayPoints()` ?

Reply:
We can do that, but why proposed check is better than present? Just curious.

Reply:
Desigion to send the next request is dependent not on the last response but from the state if any apps is subscribed. Try to make intentions, dependencies, root causes explicit for developer that is not deep in component logic. 

And reusing response from another request with replacing corr_id is `so-so` solution, if you have no other choice, it is OK, but if you can generate response this is more apropriate way.

Reply:
Done in ff6fc20e24f1ce4bcff713dc585f904332c9a812